graylog2-server/src/main/java/org/graylog2/plugin/Message.java

```diff
@@ -466,18 +455,22 @@ public boolean removeStream(Stream stream) {
         return ImmutableSet.copyOf(this.indexSets);
     }
 
+    @SuppressWarnings("unchecked")
     public List<String> getStreamIds() {
-        if (!hasField(FIELD_STREAMS)) {
-            return streams.stream().map(Stream::getId).collect(Collectors.toList());
-        }
+        List<String> streamField;
         try {
-            @SuppressWarnings("unchecked")
-            final List<String> streamIds = getFieldAs(List.class, FIELD_STREAMS);
-            return new ArrayList<>(streamIds);
+            streamField = getFieldAs(List.class, FIELD_STREAMS);
         } catch (ClassCastException e) {
             LOG.trace("Couldn't cast {} to List", FIELD_STREAMS, e);
-            return Collections.emptyList();
+            streamField = Collections.emptyList();
+        }
+
+        final Set<String> streamIds = streamField == null ? new HashSet<>(streams.size()) : new HashSet<>(streamField);
+        for (Stream stream : streams) {
+            streamIds.add(stream.getId());
         }
+
+        return new ArrayList<>(streamIds);
```

Comment:
I'm totally *for* changing the signature of this method, even if it's breaking the API in this case.